### DISCUSSION TITLE
Shortcut to toggle whitespace rendering on/off?

### DISCUSSION THREAD

**Discussion by pieqq** [helix-editor/helix#7908] [Q&A] [answered]
Hello!

Sometimes, I want to check that my files don't include unnecessary invisible characters. Helix has a command for that: `:set whitespace.render all`. To disable it, I can run `:set whitespace.render none`.

Is it possible to have a shortcut to **toggle** this option? So far, the only option I found was to have two separate shortcuts: one to render all the whitespaces, one to render none, but this is not great.

So, is it possible to somehow have a shortcut that, when pressed, would do something like:

```
if whitespace.render == all:
    whitespace.render == none
else:
    whitespace.render == all
```

**Comment by CptPotato** [comment]
Looks like this isn't supported, yet.

The command should be `:toggle whitespace.render all none`, but helix reports `Configuration whitespace.render does not support toggle yet`.

**Reply by David-Else** [reply]
It is supported now, try with the latest master! :)

Bonus zen mode config:

```
z = ":toggle gutters.line-numbers.min-width 48 3"
```

**Reply by CptPotato** [reply]
Yup, I am using master. Toggle is specifically not working with `whitespace.render` currently.

**Reply by David-Else** [reply, marked as answer]
```
[keys.normal.space]
x = ":toggle whitespace.render all none"
```
Works perfectly here, it may have been made possible by: https://github.com/helix-editor/helix/pull/7877

Maybe it is your evil Windows ways causing problems :)

**Reply by pieqq** [reply]
Thank you for this quick answer! I'll wait for the next release then :)

**Reply by fintara** [reply]
For other people hitting "Configuration whitespace.render does not support toggle yet": if you have `[editor.whitespace.render]` in your config file with individual space/tab/etc, then the toggling will show this error. Remove the whole section and the toggling will work fine.

**Reply by GerwazyMiod** [reply]
> For other people hitting "Configuration whitespace.render does not support toggle yet": if you have [editor.whitespace.render] in your config file with individual space/tab/etc, then the toggling will show this error. Remove the whole section and the toggling will work fine.

This also applies to `set` commands - bindings:
```
F1 = ":set whitespace.render none"
F2 = ":set whitespace.render all"
```
wasn't working when `[editor.whitespace.render]` was present in config

**Reply by Axlefublr** [reply]
I would love for helix to support having some whitespace always visible, and other ones sometimes visible, via toggles

**Reply by pupto-bismol** [reply]
I was fiddling with this exact issue, and from the docs I saw that `:toggle` will actually take in JSON values if appropriate. So, you can actually write something like:

```
:toggle whitespace.render {"space":"none"} {"space":"all"}
```

Unfortunately, for some reason running this command will only ever output `'whitespace.render' is now set to {"space":"none"}`, and it will never toggle back to "all." This is true whether or not you run it in the command line or set a keybinding. And, if you swap the order of the JSON arguments in the code block above, it will do the exact same thing but then *only* ever set it to "all" instead. I tried to set every value possible at once to see if that would work, i.e:

```
:toggle whitespace.render {
  "space":"none",
  "nbsp":"none",
  "nnbsp":"none",
  "tab":"none",
  "newline":"none",
  "tabpad":"none"
} {
  "space":"all",
  "nbsp":"all",
  "nnbsp":"all",
  "tab":"all",
  "newline":"all",
  "tabpad":"all"
}
```

but this also did not work. I'm not sure if this is a bug or just some quirk of the toggle behaviour, but either way it does not seem intended.

**Reply by 8ullyMaguire** [reply]
> ```
> [keys.normal.space]
> x = ":toggle whitespace.render all none"
> ```
> 
> Works perfectly here, it may have been made possible by: #7877
> 
> Maybe it is your evil Windows ways causing problems :)

You were pressing <kbd>x</kbd> and only selecting a line. You have to press <kbd>Space</kbd> + <kbd>x</kbd>.
